File: src/main/java/com/codingagent/llm/Message.java
package com.codingagent.llm;

import java.util.List;
import java.util.Map;

/**
 * 所有 LLM 提供商共用的统一消息类型。
 * SYSTEM 消息携带指令，USER/ASSISTANT 携带对话，
 * TOOL 消息通过 {@code toolCallId} 关联对应的工具调用结果。
 * {@code reasoningContent} 用于 DeepSeek 等模型的思考模式。
 */
public record Message(Role role, String content, List<ToolCall> toolCalls,
                      String toolCallId, String reasoningContent) {

    public enum Role { SYSTEM, USER, ASSISTANT, TOOL }

    // -- 向后兼容构造器 --
    public Message(Role role, String content, List<ToolCall> toolCalls, String toolCallId) {
        this(role, content, toolCalls, toolCallId, null);
    }

    public static Message system(String content) {
        return new Message(Role.SYSTEM, content, null, null, null);
    }

    public static Message user(String content) {
        return new Message(Role.USER, content, null, null, null);
    }

    public static Message assistant(String content) {
        return new Message(Role.ASSISTANT, content, null, null, null);
    }

    public static Message assistant(String content, String reasoningContent) {
        return new Message(Role.ASSISTANT, content, null, null, reasoningContent);
    }

    /** 携带工具调用的 assistant 消息。压缩器保证与 toolResult 成对不拆分。 */
    public static Message assistantWithToolCalls(ToolCall toolCall) {
        return new Message(Role.ASSISTANT, null, List.of(toolCall), null, null);
    }

    /** 携带工具调用和思考内容的 assistant 消息。 */
    public static Message assistantWithToolCalls(ToolCall toolCall, String reasoningContent) {
        return new Message(Role.ASSISTANT, null, List.of(toolCall), null, reasoningContent);
    }

    /** 工具结果消息，通过 toolCallId 与对应的 assistant 消息配对。 */
    public static Message toolResult(String toolCallId, String content) {
        return new Message(Role.TOOL, content, null, toolCallId, null);
    }

    /** assistant 消息中的单次工具调用。 */
    public record ToolCall(String id, String name, Map<String, Object> arguments) {}
}

File: internal/llm/message.go
package llm

import "maps"

// Role represents the role of a message in a conversation.
type Role string

const (
	RoleSystem    Role = "system"
	RoleUser      Role = "user"
	RoleAssistant Role = "assistant"
	RoleTool      Role = "tool"
)

// Message is the unified message type shared across all LLM providers.
type Message struct {
	Role             Role
	Content          string
	ToolCalls        []ToolCall
	ToolCallID       string
	ReasoningContent string // DeepSeek thinking mode passthrough
}

// ToolCall represents a single tool invocation within an assistant message.
type ToolCall struct {
	ID        string
	Name      string
	Arguments map[string]any
}

// System creates a system message.
func System(content string) Message {
	return Message{Role: RoleSystem, Content: content}
}

// User creates a user message.
func User(content string) Message {
	return Message{Role: RoleUser, Content: content}
}

// Assistant creates a plain assistant text message.
func Assistant(content string) Message {
	return Message{Role: RoleAssistant, Content: content}
}

// AssistantWithReasoning creates an assistant message with reasoning content (DeepSeek).
func AssistantWithReasoning(content, reasoning string) Message {
	return Message{Role: RoleAssistant, Content: content, ReasoningContent: reasoning}
}

// AssistantWithToolCalls creates an assistant message carrying tool calls.
func AssistantWithToolCalls(tc ToolCall) Message {
	return Message{Role: RoleAssistant, ToolCalls: []ToolCall{tc}}
}

// AssistantWithToolCallsAndReasoning creates an assistant message with tool calls and reasoning.
func AssistantWithToolCallsAndReasoning(tc ToolCall, reasoning string) Message {
	return Message{Role: RoleAssistant, ToolCalls: []ToolCall{tc}, ReasoningContent: reasoning}
}

// ToolResult creates a tool result message paired with its tool call via ID.
func ToolResult(toolCallID, content string) Message {
	return Message{Role: RoleTool, Content: content, ToolCallID: toolCallID}
}

// CloneToolCall returns a deep copy of a ToolCall (arguments map is copied).
func CloneToolCall(tc ToolCall) ToolCall {
	cloned := ToolCall{ID: tc.ID, Name: tc.Name}
	if tc.Arguments != nil {
		cloned.Arguments = maps.Clone(tc.Arguments)
	}
	return cloned
}
